### ISSUE TITLE
Buffering in memory when storing on S3

### ISSUE THREAD

**Issue by Wizarth** [strongloop/loopback-component-storage#116]
Hello,

I have a Loopback system using Amazon provider for storage. I'm running into an issue where the Loopback based server appears to be buffering the incoming data in memory, while streaming to S3. This is a problem when the source of the data is LAN local servers, and the incoming data rate exceeds the outgoing data rate. In particular, we are generating very large files (up to GB's), and the outgoing rate to S3 appears to be limited to 3MB/s or so. NodeJS/Loopback will consume all the memory in the system (up to approx 3GB, given the other things on the system) before being killed by the kernel.

In testing, if I POST to the Loopback server (using curl) with rate limited to 3MB/s, from one of the LAN machines, then there is no spike in memory usage on the Loopback server.

Ideally, I'd like to see it that the incoming data stream is throttled (back pressure?) when the outgoing data rate is too low.

Since the Streams interface appears to have changed many times over the history of NodeJS, I've tested this with 0.10, 0.12, and 4.4.0 . All these versions showed the same issue.

Please let me know if there's any additional information needed.

**Comment by richardpringle** [comment]
@Wizarth, thanks for the detailed description!

I am putting a feature request label on this issue so it can be considered for the next release of this module. 

If you would like to see this change sooner rather than later... you can always try to implement it yourself and submit a PR! We're relying heavily on community contribution on our extended modules  right now as we are trying to get a new release of the LoopBack core out.

**Comment by Wizarth** [comment]
I've done some more digging on this, and it looks like the issue is in storage-handler.js and how it interacts with Formidable.

In exports.upload, form.handlePart pipe()s the part into the writer. However, in Formidable, Part is only a Stream (not ReadStream) and doesn't have any ability to handle pause (see IncomingForm._initMultipart).

Looking at the implementation of IncomingForm.handlePart in Formidable, rather than calling pipe to a file, they handle data events on the part. There they call pause on the Incoming form, call WriteStream.write from the data events, then call resume on the form again in the callback.

Arguably this is an issue in Formidable, using Stream rather than ReadStream (and I might forward this to them as well) but I figured I'd document what's happening here since I've already made this issue.
